### PR TITLE
fix (server): Better error handling in broadcast signal in Alfred

### PR DESCRIPTION
**Pull request by ghoshkaj**
The [first version of the business logic](https://github.com/microsoft/FluidFramework/pull/16451) was very fragile and caused Alfred to crash on malformed JSON. This PR moves the JSON enforcement logic out to the router and makes Alfred less fragile internally by logging instead of throwing errors from the event emitter.

### CODE REVIEW

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/api.ts [hunk 1]
```diff
@@ -144,14 +145,24 @@ export function create(
...
 			const documentId = getParam(request.params, "id");
-			const signalContent = getParam(request.body, "singalContent");
 			try {
-				const signalRoom: IRoom = { tenantId, documentId };
-				const payload: IBroadcastSignalEventPayload = { signalRoom, signalContent };
-				collaborationSessionEventEmitter.emit("broadcastSignal", payload);
-				response.status(200).send("OK");
+				const signalContent = JSON.parse(
+					getParam(request.body, "signalContent"),
+				) as IRuntimeSignalEnvelope;
```
Comment (alexvy86) [inline server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/api.ts:151]: The `as` doesn't do any actual validation, let's add explicit checks for `typeof signalContent?.contents?.type === "string"`, `signalContents.contents.content !== undefined`. I might have gotten the paths wrong :).

Also, I think it's cleaner to limit the new try-catch block to the `JSON.parse()` and the new checks, and `return` immediately after setting the `response` properties in the catch block to avoid moving past that point (I can't recall if there's a particular pattern used in Express routes for that or a simple `return;` is enough).

File: server/routerlicious/packages/lambdas/src/alfred/index.ts [hunk 2]
```diff
@@ -46,6 +46,7 @@ import {
 	generateClientId,
 	IRuntimeSignalEnvelope,
 } from "../utils";
+export { IRuntimeSignalEnvelope } from "../utils";
```
Comment (alexvy86) [inline server/routerlicious/packages/lambdas/src/alfred/index.ts:49]: This can be removed and index can export it directly from "./utils" instead of from "./alfred".

File: server/routerlicious/packages/lambdas/src/alfred/index.ts [hunk 3]
```diff
@@ -648,15 +650,15 @@ export function configureWebSocketServices(
...
-							);
 						}
+					} catch (error) {
+						const errorMsg = `Broadcast-signal failed with: ${error}`;
+						return handleServerError(
+							logger,
+							errorMsg,
+							claims.documentId,
+							claims.tenantId,
+						);
```
Comment (alexvy86) [inline server/routerlicious/packages/lambdas/src/alfred/index.ts:661]: We need to replace the `return` here with just a Lumberjack logging, right? And I think we can pass the error object directly to that call, so the additional message can just be "Unexpected error handling broadcastSignal event".

File: server/routerlicious/packages/lambdas/src/alfred/index.ts [hunk 4]
```diff
@@ -622,18 +623,19 @@ export function configureWebSocketServices(
...
-						signalRoom.documentId === room.documentId &&
-						signalRoom.tenantId === room.tenantId
-					) {
-						try {
-							const contents = JSON.parse(signalContent) as IRuntimeSignalEnvelope;
-							const runtimeMessage = createRuntimeMessage(contents);
+					try {
+						const signalRoom = broadcastSignal.signalRoom;
+						const signalContent =
+							broadcastSignal.signalContent as IRuntimeSignalEnvelope;
```
Comment (Josmithr) [inline server/routerlicious/packages/lambdas/src/alfred/index.ts:629]: Nit: I generally recommend casting to `Partial<Foo>` whenever casting an object of unknown type. That way you're forced to handle the cases where the data you're looking for isn't there. Doesn't look like it would have a major impact on this use case, but figured I would call it out 🙂

Comment (Josmithr) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts:629]: Also, what does it mean if we get data that isn't `IRuntimeSignalEnvelope`? Is that invalid (should we throw / log an error)? Or does it mean something else? Would probably be worth documenting.

File: server/routerlicious/packages/lambdas/src/alfred/interfaces.ts [hunk 5]
```diff
@@ -56,7 +56,7 @@ export interface IBroadcastSignalEventPayload {
 	/**
 	 * Content of the runtime signal introduced from the broadcast-signal endpoint.
 	 */
-	signalContent: string;
+	signalContent: unknown;
```
Comment (Josmithr) [inline server/routerlicious/packages/lambdas/src/alfred/interfaces.ts:59]: What is the contract for the data format here? Is it defined by the signal type? These payloads are serializable, right? Should we constrain this type to something that supports serialization?

Comment (ghoshkaj) [reply server/routerlicious/packages/lambdas/src/alfred/interfaces.ts:59]: The contract is a stringified version of IRuntimeSignalEnvelope. @alexvy86 why did we change this to unknown again? I can't remember the exact reason -- for a while I was mistakenly using a mixture of JSON and stringified JSON, but the final (correct) version looks like this:

```
curl -X POST -H "Content-Type: application/json" -d '{"signalContent":"{\"contents\":{\"type\":\"ExternalDataChanged_V1.0.0\", \"content\":{\"taskListId\":\"task-list-1\"}}}"}' localhost:7070/tinylicious/76e92769-295e-46d2-8fe9-02aa0ea9447e/broadcast-signal
```

Comment (ghoshkaj) [reply server/routerlicious/packages/lambdas/src/alfred/interfaces.ts:59]: Also the route is explicitly throwing if it's not in that type now.

Comment (alexvy86) [reply server/routerlicious/packages/lambdas/src/alfred/interfaces.ts:59]: Hmm the change to `unknown` was supposedly to avoid the need for the caller of the API to provide it as a string, and instead provide it as an actual JSON object, no?

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/api.ts [hunk 6]
```diff
@@ -145,24 +145,24 @@ export function create(
...
-				try {
-					const signalRoom: IRoom = { tenantId, documentId };
-					const payload: IBroadcastSignalEventPayload = { signalRoom, signalContent };
-					collaborationSessionEventEmitter.emit("broadcastSignal", payload);
-					response.status(200).send("OK");
-				} catch (error) {
-					response.status(500).send(error);
-				}
-			} catch (error) {
+			const signalContent = JSON.parse(
```
Comment (alexvy86) [inline server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/api.ts:148]: As mentioned in [this other comment thread](https://github.com/microsoft/FluidFramework/pull/17540#discussion_r1343243084), I think this should go away if we let `signalContent` be an actual JSON object instead of a string. If we don't go that route though, this should be try-catch-ed and return a 400 if the call to `JSON.parse()` fails.

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/api.ts [hunk 7]
```diff
@@ -145,24 +145,24 @@ export function create(
...
+			const validSignalContent =
+				typeof signalContent?.contents?.type === "string" &&
+				signalContent?.contents?.content !== undefined;
+			if (!validSignalContent) {
 				response
 					.status(400)
-					.send(
-						`signalContent should contain 'content' and 'type' keys. Error: ${error}`,
-					);
+					.send(`signalContent should contain 'content' and 'type' keys.`);
```
Comment (alexvy86) [inline server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/api.ts:157]: This needs a `return;` right after it, `.send()` won't stop the execution of the method so after this is called the `try-catch` after the `if` will still execute.

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/api.ts [hunk 8]
```diff
@@ -145,24 +145,24 @@ export function create(
...
+			}
+			try {
+				const signalRoom: IRoom = { tenantId, documentId };
+				const payload: IBroadcastSignalEventPayload = { signalRoom, signalContent };
+				collaborationSessionEventEmitter.emit("broadcastSignal", payload);
+				response.status(200).send("OK");
+			} catch (error) {
+				response.status(500).send(error);
```
Comment (alexvy86) [inline server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/api.ts:165]: Note: here it's not an issue to not have a return because these are the end of the method already, but it's still true that they don't immediately make the handler stop.

File: server/routerlicious/packages/lambdas/src/alfred/interfaces.ts (hunk 5 above)
Comment (Josmithr) [reply server/routerlicious/packages/lambdas/src/alfred/interfaces.ts:59]: Should this be typed as a `Record`, then?

File: server/routerlicious/packages/lambdas/src/alfred/index.ts (hunk 4 above)
Comment (Josmithr) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts:628]: > In this case would it be Partial?

Only as a means of guarding against malformed input. By typing it as `Partial<IRuntimeSignalEnvelope>`, the type system will force you to validate that the fields you care about are actually there.

File: server/routerlicious/packages/lambdas/src/alfred/interfaces.ts (hunk 5 above)
Comment (alexvy86) [reply server/routerlicious/packages/lambdas/src/alfred/interfaces.ts:59]: Yeah, that might actually be a better option.

Comment (Josmithr) [reply server/routerlicious/packages/lambdas/src/alfred/interfaces.ts:59]: We should also be sure to document the expectation that this is a JSONified version of `IRuntimeSignalEnvelope`.

Comment (ghoshkaj) [reply server/routerlicious/packages/lambdas/src/alfred/interfaces.ts]: Okay, can you take a look at the latest? I still don't quite understand how to use Record -- should I be defining the IRuntimeSignalEnvelope interface as a Record? Also, the Partial<IRuntimeSignalEnvelope> forces me to also make the definition a partial -- is that what we want?

Comment (Josmithr) [reply server/routerlicious/packages/lambdas/src/alfred/interfaces.ts]: Actually, sorry, `Record` may or may not be the best option here. I guess my first question is: should this just be typed as `IRuntimeSignalEnvelope`, if that's the only value we support? Or am I misunderstanding?

If not that, then maybe using our shared `Jsonable` type would be best? It handles recursive cases better than `Record` would.

Comment (Josmithr) [reply server/routerlicious/packages/lambdas/src/alfred/interfaces.ts]: > Also, the Partial forces me to also make the definition a partial -- is that what we want?

`Partial` is a type built into TypeScript. It defines a type that is the same as the parameter you give it, except its members are optional.

E.g. 
```typescript
interface Foo {
  bar: string;
  baz: () => number;
}

/**
 * Equivalent to 
 * {
 *   bar?: string;
 *   baz?: () => number;
 * }
 */
type PartialFoo = Partial<Foo>;
```

Comment (ghoshkaj) [reply server/routerlicious/packages/lambdas/src/alfred/interfaces.ts]: Got it! Okay so I went ahead and used `IRuntimeSignalEnvelope` here. The only other thing to address I think now is converting it on receipt inside Alfred: https://github.com/microsoft/FluidFramework/pull/17540/files#diff-56222e558ec8d0b289c0be5a65d798a1ec047f1af4ad33aab0c5454e38381f22R148-R151
Screenshot 2023-10-04 at 5 04 49 PM

I would like to prevent using that `as unknown as IRuntimSignalEnvelope` on line 151, but if I remove the `as unknown`  and cast it directly to IRuntimSignalEnvelope, it errors with:

``` Conversion of type 'string' to type 'IRuntimeSignalEnvelope' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.ts(2352)```

Comment (ghoshkaj) [reply server/routerlicious/packages/lambdas/src/alfred/interfaces.ts]: Ohh I just realized what the problem is -- the function `getParam` requires the second param to be a string, but it isn't anymore, since it's a JSON. Hrm.

Comment (ghoshkaj) [reply server/routerlicious/packages/lambdas/src/alfred/interfaces.ts]: Okay, I'm grabbing the signalContent directly from the body now, and casting it directly as IRuntimSignalEnvelope. It should all work, please take a look! I wonder how to document now that the body will have to be JSON though :/ ANd do we do a header check?

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/api.ts [hunk 9]
```diff
@@ -144,14 +145,22 @@ export function create(
...
+					.status(400)
+					.send(`signalContent should contain 'content' and 'type' keys.`);
+				return;
+			}
 			try {
 				const signalRoom: IRoom = { tenantId, documentId };
 				const payload: IBroadcastSignalEventPayload = { signalRoom, signalContent };
 				collaborationSessionEventEmitter.emit("broadcastSignal", payload);
 				response.status(200).send("OK");
+				return;
```
Comment (alexvy86) [inline server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/api.ts:160]: As we discussed offline, the `collaborationSessionEventEmitter.emit()` call a few lines above this needs a check for `undefined`.

File: server/routerlicious/packages/routerlicious-base/src/test/alfred/api.spec.ts [hunk 10]
```diff
@@ -378,10 +385,46 @@ describe("Routerlicious", () => {
...
 						defaultDocumentDeleteService,
+						null,
+						null,
+						defaultCollaborationSessionEventEmitter,
 					);
 					supertest = request(app);
 				});
 
+				describe("api/v1", () => {
+					it("/api/v1/:tenantId/:id/broadcast-signal 200", async () => {
```
Comment (alexvy86) [inline server/routerlicious/packages/routerlicious-base/src/test/alfred/api.spec.ts:396]: And we can name the next test `"Invalid request content"`.
```suggestion
				describe("/api/v1/:tenantId/:id/broadcast-signal", () => {
					it("Successful request", async () => {
```

File: server/routerlicious/packages/routerlicious-base/src/test/types/validateServerRouterliciousBasePrevious.generated.ts
Comment (alexvy86) [inline server/routerlicious/packages/routerlicious-base/src/test/types/validateServerRouterliciousBasePrevious.generated.ts:1]: I believe the typetest changes are because of recent improvements to our type test generation which are probably just hitting server now. Let's do a separate PR that does nothing but update the typetests in server, so this change goes there, and this PR doesn't include changes unrelated to its purpose.

Comment (ghoshkaj) [reply server/routerlicious/packages/routerlicious-base/src/test/types/validateServerRouterliciousBasePrevious.generated.ts:1]: Looks like this and the other .generate file are already up to do date in main. Just udpated with main so these changes are gone defacto!

**Review by alexvy86 — APPROVED** [review]
🚀
